### PR TITLE
ui/eth: wallet setup hints

### PR DESCRIPTION
**Pull request by norwnd**
Resolves https://github.com/decred/dcrdex/issues/2245, https://github.com/decred/dcrdex/issues/2246.

image

There are 2 providers now there ^ (with 2nd commit) actually, screenshot taken based on 1st commit from this PR).

### CODE REVIEW

File: client/asset/eth/eth.go [hunk 1]
```diff
@@ -121,6 +121,17 @@ var (
 			Repeatable: providerDelimiter,
 			Required:   true,
 		},
+		// Allow for single Provider setup, still nag the user to specify 2nd one for
+		// additional safety in case 1st provider fails during trade settlement.
+		{
+			Key:         providersKey,
+			DisplayName: "Provider",
+			Description: "Specify one or more providers. For infrastructure " +
+				"providers, use an https address. Only url-based authentication " +
```
Comment (chappjc) [inline client/asset/eth/eth.go:130]: "use an https address"?  We're supposed to be recommending wss

File: client/webserver/site/src/html/forms.tmpl [hunk 2]
```diff
@@ -1,4 +1,7 @@
 {{define "walletConfigTemplates"}}
+<div id="walletCfgGuide" class="py-2 d-hide">
+  Make sure you follow:  <a id="walletCfgGuideLink" target="_blank" class="py-2 hyperlink">configuration guide</a><br>
```
Comment (chappjc) [inline client/webserver/site/src/html/forms.tmpl:3]: Any text like this unfortunately needs to go in the locales.js like `[[[wallet_follow_the_guide]]]`

Regarding the text itself, why the colon?  Wouldn't it read better as "Make sure you follow the configuration guide."

**Review by chappjc — COMMENTED** [review]
@buck54321 Do you think we need the guide link in the backend as part of the wallet definition?  I don't see a reason why the frontend shouldn't just have some map of guide links per asset, similar to how we do the explorer links.

File: client/webserver/site/src/html/forms.tmpl (hunk 2 above)
Comment (buck54321) [reply client/webserver/site/src/html/forms.tmpl:3]: Better with no colon.

**Review by buck54321 — CHANGES_REQUESTED** [review]
Please consider these changes.
https://github.com/buck54321/dcrdex/commit/a58ab2b544b767303d692e95435fc3f4ae47dcfd
image

**Comment by buck54321** [comment]
> @buck54321 Do you think we need the guide link in the backend as part of the wallet definition? I don't see a reason why the frontend shouldn't just have some map of guide links per asset, similar to how we do the explorer links.

I had made this recommendation and I stand by it. Storing it in the front end code is hacky, and precludes non-UI consumers from accessing the information.

**Comment by chappjc** [comment]
> Storing it in the front end code is hacky, and precludes non-UI consumers from accessing the information.

Alright, but I don't get why it's hacky, or why non-UI consumers would access a docs URL.  Filling up the backend code with this makes me cringe because it is in no way required to make or even configure the backend -- it's entirely docs i.e. UI.

**Comment by norwnd** [comment]
Upon retesting I realised 2 providers by default doesn't quite work and needs further adjustments:

[add-2nd-provider-not-working.webm](https://user-images.githubusercontent.com/112318969/229025783-53871df6-d2b0-4fa1-8499-a0c1d1979e7e.webm)

possible way to resolve it: https://github.com/decred/dcrdex/pull/2261/commits/2f0fc860acf53e986b9c75a95b189026a40bd11c

File: client/asset/eth/eth.go (hunk 1 above)
Comment (norwnd) [reply client/asset/eth/eth.go]: Changed this text a bit.

**Comment by norwnd** [comment]
> Please consider these changes.
https://github.com/buck54321/dcrdex/commit/a58ab2b544b767303d692e95435fc3f4ae47dcfd

It seems these need to be adjusted such that they also apply to "Reconfigure" form ? Basically similar how I have it now.

**Review by chappjc — APPROVED** [review]
Looks good to me.  I personally don't care for either `RepeatN` or `GuideLink` in the `ConfigOpts` defined on the backend, but I'm overruled.  It's certainly wrong to just duplicate the `providersKey` config opt entry, so `RepeatN` is right in that regard.  I think both the guide link and duplicating a form entry are entirely frontend concerns.  A headless bot wouldn't have any use for these config opt fields.

**Comment by buck54321** [comment]
> I think both the guide link and duplicating a form entry are entirely frontend concerns. A headless bot wouldn't have any use for these config opt fields.

What about RPC users? They need to know wallet configuration instructions too.

File: client/webserver/site/src/js/forms.ts [hunk 3]
```diff
@@ -557,12 +562,18 @@ export class WalletConfigForm {
   /*
    * update creates the dynamic form.
    */
-  update (configOpts: ConfigOption[] | null, activeOrders: boolean) {
+  update (guideLink: string, configOpts: ConfigOption[] | null, activeOrders: boolean) {
     this.assetHasActiveOrders = activeOrders
     this.configElements = []
     this.configOpts = configOpts || []
     Doc.empty(this.dynamicOpts, this.defaultSettings, this.loadedSettings)
```
Comment (buck54321) [inline client/webserver/site/src/js/forms.ts:569]: As I [suggested before](https://github.com/decred/dcrdex/pull/2261#pullrequestreview-1359740617), please move the `walletCfgGuide` out of the `walletSettings` and into the `mainForm`. This is not the purview of the dynamic form section.

**Comment by chappjc** [comment]
> > I think both the guide link and duplicating a form entry are entirely frontend concerns. A headless bot wouldn't have any use for these config opt fields.
> 
> What about RPC users? They need to know wallet configuration instructions too.

These config options are not available via the RPC server.  The `newwallet` syntax is just to specify the settings:

```
settings (string): A JSON-encoded string->string mapping of additional
       configuration settings. These settings take precedence over any settings
       parsed from file. e.g. '{"account":"default"}' for Decred accounts
...
```

While that could be changed, and a walletdefinitions RPC added, I don't think it's worth it.

**Comment by buck54321** [comment]
Huh. I thought you could access `SupportedAsset`s through RPC. Regardless, I don't view this information as being specific to our user interface implementation. It's asset-related information that could be useful to other users of `Core`. `Core` is not only for headless bots. More than anything though, I don't think this is worth arguing about any more. If it's important to you to move it to the front end, that's fine.

`RepeatN`, on the other hand, **is** extremely specific to our UI implementation, but I can see an argument for keeping it with the `ConfigOption` parameters too, so it's whatever. I don't want to keep dragging @norwnd into this philosophical debate. Let's just move forward.

**Comment by chappjc** [comment]
> If it's important to you to move it to the front end, that's fine.

No, let's move on.

**Comment by norwnd** [comment]
>  I don't want to keep dragging @norwnd into this philosophical debate. Let's just move forward.

Yeah, I don't think it makes much difference currently whether it's defined on backend or front end.

File: client/webserver/site/src/js/forms.ts [hunk 4]
```diff
@@ -67,7 +67,7 @@ export class NewWalletForm {
   current: CurrentAsset
   pwHiders: HTMLElement[]
   subform: WalletConfigForm
-  currentWalletType: string
```
Comment (norwnd) [inline client/webserver/site/src/js/forms.ts:70]: Removed `currentWalletType` since unused.

File: client/webserver/site/src/js/forms.ts (hunk 3 above)
Comment (norwnd) [reply client/webserver/site/src/js/forms.ts]: Moved now to a separate `template` that gets embedded into **both** `mainForm` and `reconfigForm` https://github.com/decred/dcrdex/pull/2261/commits/26c65ba6e867c76f3a7e0d1c77dee5cf11ac0aa8, see also [why I couldn't apply your change as is](https://github.com/decred/dcrdex/pull/2261#issuecomment-1491302624).

But some JS code is duplicated now (to serve 2 different forms), not sure if you'd want to dedup it by introducing additional entity that combines settings+guide into single section or something like that, so leaving like that for now.

Comment (buck54321) [reply client/webserver/site/src/js/forms.ts]: Sorry. I see what you're saying now. Less duplication to put it in the already templated section. The new arguments to `update` stuck out to me as a smell though. You can put it back if you want, but I'm fine either way.

**Review by buck54321 — CHANGES_REQUESTED** [review]
This looks good, but I think I found some improvements. Lemme know what you think.
https://github.com/buck54321/dcrdex/compare/26c65ba...buck54321:dcrdex:repeat-addopt-internal
1. The `repeatableCounter` business was apparently only to avoid duplicate IDs for javascript `id => from` attribute matching. That's simplified.
2. Handling of `RepeatN` is internal to `addOpt`, reducing duplication
3. An extra blank field(s) is added to the reconfig form

File: client/webserver/site/src/js/forms.ts (hunk 3 above)
Comment (norwnd) [reply client/webserver/site/src/js/forms.ts]: I think it's fine to leave it like it currently is for now.

**Comment by norwnd** [comment]
> This looks good, but I think I found some improvements. Lemme know what you think. [buck54321/dcrdex@26c65ba...buck54321:dcrdex:repeat-addopt-internal](https://github.com/buck54321/dcrdex/compare/26c65ba...buck54321:dcrdex:repeat-addopt-internal)
> 
> 1. The `repeatableCounter` business was apparently only to avoid duplicate IDs for javascript `id => from` attribute matching. That's simplified.
> 2. Handling of `RepeatN` is internal to `addOpt`, reducing duplication
> 3. An extra blank field(s) is added to the reconfig form

Yep ty, looks cleaner and seems to be working, applied https://github.com/decred/dcrdex/pull/2261/commits/f0068475323aad5d40fcb8de9590322c714d8860

**Comment by norwnd** [comment]
And now rebased.

**Comment by chappjc** [comment]
Would like to ship this in the coming patch that adds zcash 5.4.2 support and dgb

**Review by buck54321 — APPROVED** [review]

**Comment by chappjc** [comment]
Thanks.  `GuideLink` will be nice the Electrum wallet's too.  I'll get around to making a wiki page for that showing screens like [this](https://github.com/decred/dcrdex/wiki/images/electrum-rpc-config.png)

**Review by chappjc — APPROVED** [review]

**Pull request merged**